Dev

RSWilson1 opened the pull request:
- Added unittests for code
- Refactoring of gene_2annotation.py
- PEP8 changes and formatting

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)
<!-- Reviewable:end -->

RSWilson1 committed:
Refactored, for better error handling and testing.

RSWilson1 committed:
small refactoring out of main function

RSWilson1 committed:
Added testing for stout

RSWilson1 committed:
formatting def line

RSWilson1 committed:
IGV report doc string

RSWilson1 committed:
Writing unittests and refactoring code for better testing
- changed handling of chromosome
- moved into classes
- used patching to mock faidx

RSWilson1 committed:
changed to be compliant with test

RSWilson1 committed:
added some extensions to ignore

RSWilson1 committed:
PEP8

pep8speaks commented:
Hello @RSWilson1! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`construct_vcf.py`](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py):

> [Line 80:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py#L80): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 85:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 92:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py#L92): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 105:13](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py#L105): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'
> [Line 201:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py#L201): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 203:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/construct_vcf.py#L203): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

* In the file [`gene_annotation2bed.py`](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/gene_annotation2bed.py):

> [Line 287:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/gene_annotation2bed.py#L287): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 290:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/gene_annotation2bed.py#L290): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 321:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/gene_annotation2bed.py#L321): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)



* In the file [`tests/test_construct_vcf.py`](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py):

> [Line 29:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L29): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (160 > 79 characters)
> [Line 54:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 63:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L63): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 100:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L100): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 101:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 109:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L109): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 118:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L118): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 164:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L164): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 183:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_construct_vcf.py#L183): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)

* In the file [`tests/test_gene_annotation2bed.py`](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py):

> [Line 176:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py#L176): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 202:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py#L202): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 209:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py#L209): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 221:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py#L221): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 230:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py#L230): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 246:80](https://github.com/eastgenomics/gene_annotation2bed/blob/35cd5043c504ac7340c557075765b57f9ace66f2/tests/test_gene_annotation2bed.py#L246): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)



##### Comment last updated at 2024-01-09 16:21:44 UTC

RSWilson1 committed:
updated ignore

RSWilson1 committed:
PEP8

RSWilson1 committed:
Added example files for tests and updated ignore to include them.

RSWilson1 committed:
updated ignore

RSWilson1 committed:
Updated docstrings to make tests clearer.

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 0 of 8 files reviewed, 2 unresolved discussions

___
*[`tests/test_construct_vcf.py` line 11 at r2](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1EisJDQ3IfNbdFIyO:-Nm1EisJDQ3IfNbdFIyP:bkr90yb) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/727798ab47b98adca13e3a72b00e9197f8ec0d96/tests/test_construct_vcf.py#L11)):*
> ```Python
> 
> 
> class TestConstructVCF(unittest.TestCase):
> ```

Should I break my tests into individual classes?
___
*[`tests/test_construct_vcf.py` line 14 at r2](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1EobmALwGphDnh5c7:-Nm1EobmALwGphDnh5c8:b-37yyuy) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/727798ab47b98adca13e3a72b00e9197f8ec0d96/tests/test_construct_vcf.py#L14)):*
> ```Python
>     def setUp(self):
>         # Define paths for sample data
>         self.bed_file_path = f"{TEST_DATA_DIR}/example_bed_hg38.bed"
> ```

Is a setUp the best place for these?


<!-- Sent from Reviewable.io -->

jethror1 reviewed (changes requested):
> Reviewed 1 of 6 files at r1, 1 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 2 of 8 files reviewed, 10 unresolved discussions (waiting on @RSWilson1)

___
*[`construct_vcf.py` line 205 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1PiDn5sT7U2f3g1A-:-Nm1PiDn5sT7U2f3g1A0:bjrq3ui) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L205)):*
> ```Python
>         output_df = pd.DataFrame(columns=columns)
> 
>         for i, row in bed_file.iterrows():
> ```

generally want to avoid using `iterrows` since its the slowest way to process anything in a df, `apply()` would be a good improvement, something like:
```
sequences = bed_file.apply(lambda x: self.fetch_nucleotides(x, self.reference_path), axis=1)
output_df = pd.concat([output_df, sequences])
```



_Code quote:_
```Python
for i, row in bed_file.iterrows():
```
___
*[`construct_vcf.py` line 213 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1QN-w5dfBtm1pXTkY:-Nm1QN-w5dfBtm1pXTkZ:b-xy5mt7) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L213)):*
> ```Python
>         # Possible new way make a list of series and then call pd.DataFrame(list_of_series)
>         # list_of_series = [pd.Series([1,2],index=cols), pd.Series([3,4],index=cols)]
>         # df = pd.DataFrame(list_of_series, columns=cols)
> ```

what is this for? :eyes:



_Code quote:_
```Python
        # list_of_series = [pd.Series([1,2],index=cols), pd.Series([3,4],index=cols)]
        # df = pd.DataFrame(list_of_series, columns=cols)
```
___
*[`gene_annotation2bed.py` line 315 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1RHWF8a5iGjEHVyyZ:-Nm1RHWF8a5iGjEHVyy_:bxbfc3d) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/gene_annotation2bed.py#L315)):*
> ```Python
>                      'ID': 'string', 'annotation': 'string'})
>     # Create masks for HGNC, Transcript, and Coordinates dataframes
>     try:
> ```

instead of a try except here, I think a simple assertion before would be better, i.e
```
assert 'ID' in df.columns, 'The annotation file does not contain an 'ID' column'
```



_Code quote:_
```Python
   try:
```
___
*[`tests/test_construct_vcf.py` line 11 at r2](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1EisJDQ3IfNbdFIyO:-Nm1SIyu94rojUkelUD4:b-6c6lsq) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/727798ab47b98adca13e3a72b00e9197f8ec0d96/tests/test_construct_vcf.py#L11)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

Should I break my tests into individual classes?
</blockquote></details>

possibly if you think it will make the tests more readable, I tend to as I like the separation of having 1:1 TestClass -> function, something like
```
class TestConstructVCFFetchNucleotides():
    ...


class TestConstructVCFConvertBedToVCF():
    ...
```

then the class docstring can describe more about what the function does, then each test with a docstring to explain the scenario being tested
___
*[`tests/test_construct_vcf.py` line 14 at r2](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1EobmALwGphDnh5c7:-Nm1Sjcf-wPP479sao9A:b-1uwdg9) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/727798ab47b98adca13e3a72b00e9197f8ec0d96/tests/test_construct_vcf.py#L14)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

Is a setUp the best place for these?
</blockquote></details>

looks fine to me :)
___
*[`tests/test_construct_vcf.py` line 22 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1SmMo756KTshncLR9:-Nm1SmMo756KTshncLRA:b-ghoez5) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L22)):*
> ```Python
>                                   sep="\t", header=None, names=["chr", "start",
>                                                                 "end", "info",
>                                                                 "gene"])
> ```

where you get this level of indentation I think the following is more readable:
```
self.bed_df = pd.read_csv(
    f"{TEST_DATA_DIR}/example_bed_hg38.bed",
    sep="\t",
    header=None,
    names=["chr", "start", "end", "info", "gene"]
)

```



_Code quote:_
```Python
        self.bed_df = pd.read_csv(f"{TEST_DATA_DIR}/example_bed_hg38.bed",
                                  sep="\t", header=None, names=["chr", "start",
                                                                "end", "info",
                                                                "gene"])
```
___
*[`tests/test_construct_vcf.py` line 38 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1VfScAkQ0mSw2xGP3:-Nm1VfScAkQ0mSw2xGP4:byem7la) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L38)):*
> ```Python
>         """
>         Test fetch_nucleotides method for normal chromosomes
>         Uses patch to avoid reference file dependency
> ```

can the docstring be more descriptive about the scenario you're trying to test and what it is doing?



_Code quote:_
```Python
        """
        Test fetch_nucleotides method for normal chromosomes
        Uses patch to avoid reference file dependency
```
___
*[`tests/test_construct_vcf.py` line 70 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1ToHE86IBcbQ9IStm:-Nm1ToHE86IBcbQ9IStn:b-dlx4pi) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L70)):*
> ```Python
>             self.assertEqual(result_df.iloc[1]['ALT'], 'A')
>             self.assertEqual(result_df.iloc[2]['REF'], 'C')
>             self.assertEqual(result_df.iloc[2]['ALT'], 'G')
> ```

try to avoid multiple asserts like this inside one test, as if the first fails the rest of the test won't run, some other options:
- use pytest subtest
- build a list of errors for each check and then assert that at the end
- refactor this to something like:
```
expected = [
    (result_df.iloc[0]['REF']: 'A'),
    (result_df.iloc[0]['ALT']: 'T'),
    (result_df.iloc[1]['REF']: 'T'),
    (result_df.iloc[1]['ALT']: 'A'),
    (result_df.iloc[2]['REF']: 'C'),
    (result_df.iloc[2]['ALT']: 'G')
]

assert all([x[0] == x[1] for x in expected])
```



_Code quote:_
```Python
            assert mock_faidx.call_args_list == [
                ((self.reference_path, "14:2-2"), {}),
                ((self.reference_path, "14:5-5"), {}),
                ((self.reference_path, "14:10-10"), {}),
            ]
            self.assertEqual(result_df.iloc[0]['REF'], 'A')
            self.assertEqual(result_df.iloc[0]['ALT'], 'T')
            self.assertEqual(result_df.iloc[1]['REF'], 'T')
            self.assertEqual(result_df.iloc[1]['ALT'], 'A')
            self.assertEqual(result_df.iloc[2]['REF'], 'C')
            self.assertEqual(result_df.iloc[2]['ALT'], 'G')
```
___
*[`tests/test_construct_vcf.py` line 77 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1VzF51G-_JDXfR4uy:-Nm1VzF6Br9g8XABN_5k:bjk5k3j) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L77)):*
> ```Python
>         Uses patch to avoid reference file dependency
>         """
> ```

I be here for later, need to do something else
___
*[`tests/test_construct_vcf.py` line 106 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1VcbY91OMFh7WqK3g:-Nm1VcbZ6KrCtQRpcBTN:b-9chp70) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L106)):*
> ```Python
>             self.assertEqual(result_df.iloc[1]['ALT'], 'A')
>             self.assertEqual(result_df.iloc[2]['REF'], 'C')
>             self.assertEqual(result_df.iloc[2]['ALT'], 'G')
> ```

as above



_Code quote:_
```Python
            # Assertions
            assert mock_faidx.call_args_list == [
                ((self.reference_path, "X:2-2"), {}),
                ((self.reference_path, "X:5-5"), {}),
                ((self.reference_path, "X:10-10"), {}),
            ]
            self.assertEqual(result_df.iloc[0]['REF'], 'A')
            self.assertEqual(result_df.iloc[0]['ALT'], 'T')
            self.assertEqual(result_df.iloc[1]['REF'], 'T')
            self.assertEqual(result_df.iloc[1]['ALT'], 'A')
            self.assertEqual(result_df.iloc[2]['REF'], 'C')
            self.assertEqual(result_df.iloc[2]['ALT'], 'G')
```


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
updated docstrings in igv_report.py
removed cmd not working
added download links for provenance
added refseq files for the refseq track in IGV (and updated gitignore to allow)
updated tracks with new files works for hg19 and hg38.
added zero-based for possible integration later

RSWilson1 committed:
updated to working cmd
changed argument name for clarity

RSWilson1 committed:
renamed test script

RSWilson1 committed:
rename again to correct script name that it's testing

jethror1 reviewed (changes requested):
> Reviewed 2 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 2 of 12 files reviewed, 18 unresolved discussions (waiting on @RSWilson1)

___
*[`construct_vcf.py` line 111 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAj-8CDwcyyc-pQjOs:-NmAj-8CDwcyyc-pQjOt:bz9inkq) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L111)):*
> ```Python
>             if ncbi_chr <= 22:
>                 ncbi_chr = int(chr_str)
>             else:
> ```

I think the logic could be shortened a bit here to make it clearer, and implicitly handle upper vs lowercase to be easier for the user like:

```
if chr_str.upper() == 'X':
    ncbi_str = 'X'
elif chr_str.upper() == 'Y':
    ncbi_str = 'Y'
else:
    try:
        ncbi_chr = int(chr_str)
        assert ncbi_chr <= 22
    except:
        raise ValueError(f"Error: {chr_str} is not a valid chromosome")
```

also makes your testing a bit easier :)



_Code quote:_
```Python
        if chr_str == 'X':
            ncbi_chr = 'X'
        elif chr_str == 'Y':
            ncbi_chr = 'Y'
        elif chr_str == 'y' or chr_str == 'x':
            raise ValueError(
                f"Error: {chr_str} is not a valid chromosome, use uppercase X or Y")
        else:
            try:
                ncbi_chr = int(chr_str)
            except ValueError:
                raise ValueError(f"Error: {chr_str} is not a valid chromosome")
            if ncbi_chr <= 22:
                ncbi_chr = int(chr_str)
            else:
```
___
*[`construct_vcf.py` line 141 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAidjJ5HPmQ6avgJEc:-NmAidjJ5HPmQ6avgJEd:b-k6mqbc) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L141)):*
> ```Python
>         end_nuc = end_seq.splitlines()[1]
> 
>         # # Convert the ref nucleotides to the alternate nucleotides
> ```

duplicate hash



_Code quote:_
```Python
# # C
```
___
*[`tests/test_construct_vcf.py` line 125 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAgwDh4KyxSkogxRFW:-NmAgwDh4KyxSkogxRFX:b-w2n6lk) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L125)):*
> ```Python
>             result_df = vcf_obj.fetch_nucleotides(row, self.reference_path)
>         assert str(
>             excinfo.value) == f"Error: {chromosome.replace('chr', '')} is not a valid chromosome"
> ```

to check that the expected error has the correct message, you can use the `match` parameter in `pytest.raises` like:

```
expected_error = f"Error: {chromosome.replace('chr', '')} is not a valid chromosome"

with pytest.raises(ValueError, match=expected_error):
    result_df = vcf_obj.fetch_nucleotides(row, self.reference_path)
```



_Code quote:_
```Python
        # Check error is raised:
        with pytest.raises(ValueError) as excinfo:
            result_df = vcf_obj.fetch_nucleotides(row, self.reference_path)
        assert str(
            excinfo.value) == f"Error: {chromosome.replace('chr', '')} is not a valid chromosome"
```
___
*[`tests/test_construct_vcf.py` line 143 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAhIv86W6DpR2FPsj4:-NmAhIv86W6DpR2FPsj5:b-9chp70) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L143)):*
> ```Python
>         with pytest.raises(ValueError) as excinfo:
>             result_df = vcf_obj.fetch_nucleotides(row, self.reference_path)
>         assert str(
> ```

as above



_Code quote:_
```Python
assert
```
___
*[`tests/test_construct_vcf.py` line 162 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAhRpt1azufORYmR1O:-NmAhRpt1azufORYmR1P:b-9chp70) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L162)):*
> ```Python
>         with pytest.raises(ValueError) as excinfo:
>             result_df = vcf_obj.fetch_nucleotides(row, self.reference_path)
>         assert str(
> ```

as above



_Code quote:_
```Python
assert
```
___
*[`tests/test_gene_annotation2bed.py` line 68 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAomM11QNZ-wgqzTWl:-NmAomM11QNZ-wgqzTWm:bpdcxy2) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/37069fd3aba0218a452c427bacdbe440aa8d217d/tests/test_gene_annotation2bed.py#L68)):*
> ```Python
>             "end": [14409, 25000],
>             "annotation": ["promoter_of_interest", "enhancer"],
>             "gene": ["", ""]
> ```

you've some prints in here, do you still need them?



_Code quote:_
```Python
print(input_df['Coordinates'])
```
___
*[`tests/test_gene_annotation2bed.py` line 159 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAozfrCicter8i7A4g:-NmAozfrCicter8i7A4h:b-xk0608) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L159)):*
> ```Python
> 
> 
> class Test_parsing_annotation_resource(unittest.TestCase):
> ```

can this be camel case since its a class name, also is this for `parse_annotation_tsv`? If so can you make the names match to make it easier to understand



_Code quote:_
```Python
Test_parsing_annotation_resource
```
___
*[`tests/test_gene_annotation2bed.py` line 208 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApJl6EeAXaA74aXwS:-NmApJl6EeAXaA74aXwT:b9oobm6) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L208)):*
> ```Python
>             assert transcript.empty == False  # transcript found
>             assert transcript.shape[1] == 16  # correct number of columns
>         assert coordinates_df.empty == True
> ```

as before, this is a lot of checks in one test that might fail on an earlier assert and not run the rest, consider refactoring to use subtest or build your own list of errors to check at the end (or maybe split to more tests if it seems clearer)



_Code quote:_
```Python
        self.assertEqual(print_output[0], "All rows were separated successfully")
        self.assertEqual(print_output[1], "All HGNC rows were merged successfully")
        self.assertEqual(print_output[2], "All Transcript rows were merged successfully")
        self.assertEqual(print_output[3], "No Coordinates found in the annotation file.")

        # Check df output
        assert hgnc_merged_df.empty is False
        transcripts_list = ["NM_000124", "NM_000059", "NM_000546"]
        for i in transcripts_list:
            transcript = hgnc_merged_df.loc[hgnc_merged_df['transcript_id'] == i]
            assert transcript.empty == False  # transcript found
            assert transcript.shape[1] == 16  # correct number of columns
        assert coordinates_df.empty == True
```
___
*[`tests/test_gene_annotation2bed.py` line 222 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApbyYDfZM5ilbS6gT:-NmApbyYDfZM5ilbS6gU:b6id5x1) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L222)):*
> ```Python
>         # Chr19:1-100000	Non-Oncogene
>         # chr17:1-100000	Oncogene
>         self.assertEqual(hgnc_merged_df.empty, True)
> ```

I'd be building a test data structure here to compare in one go, rather than a lot of asserts. Since it's a dataframe that's returned, you could add this correct tests data into a dict and define a df, then use `df1.equals(df2)` (just be careful of types causing issues).

e.g.:

```
correct_coordinates_df = pd.DataFrame([{'chromosome': '1', ...}])

assert correct_coordinates_df.equals(coordinates_df), "Incorrect coordinates df returned"
```


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
PR changes
- implement subtest
- group asserts together
- formatting
- added some details for docstrings

RSWilson1 committed:
rm old commented code

RSWilson1 committed:
grouped together asserts.

RSWilson1 committed:
PEP8 naming

RSWilson1 committed:
PEP8
Removing old commented lines
formatting docstring

RSWilson1 committed:
Refactored to remove multiple asserts for assering df equal

RSWilson1 committed:
PEP8

RSWilson1 committed:
PEP8

RSWilson1 committed:
PEP8

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 2 of 12 files reviewed, 17 unresolved discussions (waiting on @jethror1 and @RSWilson1)

___
*[`construct_vcf.py` line 111 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAj-8CDwcyyc-pQjOs:-NmBPsUgEevgfOwR0LAF:b-cl1qwm) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L111)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I think the logic could be shortened a bit here to make it clearer, and implicitly handle upper vs lowercase to be easier for the user like:

```
if chr_str.upper() == 'X':
    ncbi_str = 'X'
elif chr_str.upper() == 'Y':
    ncbi_str = 'Y'
else:
    try:
        ncbi_chr = int(chr_str)
        assert ncbi_chr <= 22
    except:
        raise ValueError(f"Error: {chr_str} is not a valid chromosome")
```

also makes your testing a bit easier :)
</blockquote></details>

Great yeah that saves loads of lines!
___
*[`construct_vcf.py` line 141 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAidjJ5HPmQ6avgJEc:-NmBPzJMCcBso8JNimKq:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L141)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

duplicate hash
</blockquote></details>

Done.
___
*[`construct_vcf.py` line 205 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1PiDn5sT7U2f3g1A-:-NmC7GjV5dr1nwoD7zA2:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L205)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

generally want to avoid using `iterrows` since its the slowest way to process anything in a df, `apply()` would be a good improvement, something like:
```
sequences = bed_file.apply(lambda x: self.fetch_nucleotides(x, self.reference_path), axis=1)
output_df = pd.concat([output_df, sequences])
```
</blockquote></details>

Done.
___
*[`construct_vcf.py` line 213 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1QN-w5dfBtm1pXTkY:-NmBQ0B3CYbo_M05Us1W:bf09a8r) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L213)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

what is this for? :eyes:
</blockquote></details>

Yeah this was me working on an alternative to iterrows, will remove.
___
*[`gene_annotation2bed.py` line 315 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1RHWF8a5iGjEHVyyZ:-NmC7JpI2YAMdE7oXq0X:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/gene_annotation2bed.py#L315)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

instead of a try except here, I think a simple assertion before would be better, i.e
```
assert 'ID' in df.columns, 'The annotation file does not contain an 'ID' column'
```
</blockquote></details>

Done.
___
*[`tests/test_construct_vcf.py` line 11 at r2](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1EisJDQ3IfNbdFIyO:-Nn8xUEg0iEWAGYySWq1:b-2qij8k) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/727798ab47b98adca13e3a72b00e9197f8ec0d96/tests/test_construct_vcf.py#L11)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

possibly if you think it will make the tests more readable, I tend to as I like the separation of having 1:1 TestClass -> function, something like
```
class TestConstructVCFFetchNucleotides():
    ...


class TestConstructVCFConvertBedToVCF():
    ...
```

then the class docstring can describe more about what the function does, then each test with a docstring to explain the scenario being tested
</blockquote></details>

I will look at changing this in the next PR.
___
*[`tests/test_construct_vcf.py` line 22 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1SmMo756KTshncLR9:-NmC7AgV20pR7v5tyFOK:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L22)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

where you get this level of indentation I think the following is more readable:
```
self.bed_df = pd.read_csv(
    f"{TEST_DATA_DIR}/example_bed_hg38.bed",
    sep="\t",
    header=None,
    names=["chr", "start", "end", "info", "gene"]
)

```
</blockquote></details>

Done.
___
*[`tests/test_construct_vcf.py` line 38 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1VfScAkQ0mSw2xGP3:-NmCEiA2CbBFehsOFq_l:b-dnwo9s) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L38)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

can the docstring be more descriptive about the scenario you're trying to test and what it is doing?
</blockquote></details>

updated this doc string and will repeat with the rest.
___
*[`tests/test_construct_vcf.py` line 70 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1ToHE86IBcbQ9IStm:-NmCDRaiBY77mfZvwVzx:b-nf51dn) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L70)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

try to avoid multiple asserts like this inside one test, as if the first fails the rest of the test won't run, some other options:
- use pytest subtest
- build a list of errors for each check and then assert that at the end
- refactor this to something like:
```
expected = [
    (result_df.iloc[0]['REF']: 'A'),
    (result_df.iloc[0]['ALT']: 'T'),
    (result_df.iloc[1]['REF']: 'T'),
    (result_df.iloc[1]['ALT']: 'A'),
    (result_df.iloc[2]['REF']: 'C'),
    (result_df.iloc[2]['ALT']: 'G')
]

assert all([x[0] == x[1] for x in expected])
```
</blockquote></details>

pytest subtest wasn't working very well so used a similar approach to yours.
___
*[`tests/test_construct_vcf.py` line 106 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1VcbY91OMFh7WqK3g:-NmCENmE3A2Uw72cw3Jn:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L106)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

as above
</blockquote></details>

Done.
___
*[`tests/test_construct_vcf.py` line 125 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAgwDh4KyxSkogxRFW:-NmBsyMH21lRpnW5Mwfn:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L125)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

to check that the expected error has the correct message, you can use the `match` parameter in `pytest.raises` like:

```
expected_error = f"Error: {chromosome.replace('chr', '')} is not a valid chromosome"

with pytest.raises(ValueError, match=expected_error):
    result_df = vcf_obj.fetch_nucleotides(row, self.reference_path)
```
</blockquote></details>

Done.
___
*[`tests/test_construct_vcf.py` line 143 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAhIv86W6DpR2FPsj4:-NmBsyr23hzoJAWANrZ0:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L143)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

as above
</blockquote></details>

Done.
___
*[`tests/test_construct_vcf.py` line 162 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAhRpt1azufORYmR1O:-NmBt6nX8H144bnMDysn:b-fy4y1w) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/tests/test_construct_vcf.py#L162)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

as above
</blockquote></details>

removed test as no longer needed.
___
*[`tests/test_gene_annotation2bed.py` line 68 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAomM11QNZ-wgqzTWl:-NmBsaUEAOYRCn1trC-0:b-id5bfn) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/37069fd3aba0218a452c427bacdbe440aa8d217d/tests/test_gene_annotation2bed.py#L68)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

you've some prints in here, do you still need them?
</blockquote></details>

removed all redundant prints
___
*[`tests/test_gene_annotation2bed.py` line 159 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmAozfrCicter8i7A4g:-NmCFgfT60wbxQ5JIeAi:b-896fix) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L159)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

can this be camel case since its a class name, also is this for `parse_annotation_tsv`? If so can you make the names match to make it easier to understand
</blockquote></details>

Done.
___
*[`tests/test_gene_annotation2bed.py` line 208 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApJl6EeAXaA74aXwS:-Nn8w6haDT-Q1hNuhwRc:b-7rl0jd) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L208)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

as before, this is a lot of checks in one test that might fail on an earlier assert and not run the rest, consider refactoring to use subtest or build your own list of errors to check at the end (or maybe split to more tests if it seems clearer)
</blockquote></details>

done.
___
*[`tests/test_gene_annotation2bed.py` line 222 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApbyYDfZM5ilbS6gT:-NmBLBwu7HL_kB3sp0z6:bg4l3ih) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L222)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I'd be building a test data structure here to compare in one go, rather than a lot of asserts. Since it's a dataframe that's returned, you could add this correct tests data into a dict and define a df, then use `df1.equals(df2)` (just be careful of types causing issues).

e.g.:

```
correct_coordinates_df = pd.DataFrame([{'chromosome': '1', ...}])

assert correct_coordinates_df.equals(coordinates_df), "Incorrect coordinates df returned"
```
</blockquote></details>

yeah I agree, I wrote this before working out type errors causing issues. Re-written.


<!-- Sent from Reviewable.io -->

jethror1 reviewed (changes requested):
> Reviewed 6 of 8 files at r4, 2 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 10 of 12 files reviewed, 3 unresolved discussions (waiting on @RSWilson1)

___
*[`construct_vcf.py` line 205 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1PiDn5sT7U2f3g1A-:-NnFBsxp8qq4tHGkfcWz:b-284o6v) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L205)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

is this done as the code doesn't seem to have changed? or do you want to leave it iterating?
___
*[`construct_vcf.py` line 213 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1QN-w5dfBtm1pXTkY:-NnFBzsNC0LKUQ4fNktu:b-k0v866) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L213)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

Yeah this was me working on an alternative to iterrows, will remove.
</blockquote></details>

have you not pushed changes?
___
*[`tests/test_gene_annotation2bed.py` line 208 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApJl6EeAXaA74aXwS:-NnFCau_458EZtyuIXo5:b-p9poot) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L208)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

done.
</blockquote></details>

is it?


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
Re-wrote fetch_nucleotide function.
Re-wrote iterows in convert_bed_to_vcf to use apply instead.

RSWilson1 committed:
update tests for new construct_vcf.py structure.

RSWilson1 committed:
split test into two

RSWilson1 committed:
removed old iterows code.

jethror1 reviewed (changes requested):
> Reviewed 2 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 11 of 12 files reviewed, 1 unresolved discussion (waiting on @RSWilson1)

___
*[`tests/test_gene_annotation2bed.py` line 208 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApJl6EeAXaA74aXwS:-NnivkKu7Ai5MTKMH6sV:b-5p3u0z) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L208)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

is it?
</blockquote></details>

@RSWilson1 I still think this wants changing to use subTest or similar, especially where there's assertions in a loop


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
Adding test files.
Changed test to use subtest.
All tests currently pass.

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: 11 of 15 files reviewed, 1 unresolved discussion (waiting on @jethror1)

___
*[`construct_vcf.py` line 205 at r3](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-Nm1PiDn5sT7U2f3g1A-:-Nnic0rX4CCIAGCZ9gdp:bv5s4t8) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/e75e3af94401e0dedf5d61b14bacae6b3f4ca63c/construct_vcf.py#L205)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

is this done as the code doesn't seem to have changed? or do you want to leave it iterating?
</blockquote></details>

I had an error when using apply but it's fixed now :crossed_fingers: 
___
*[`tests/test_gene_annotation2bed.py` line 208 at r4](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3#-NmApJl6EeAXaA74aXwS:-NnivDyQCn6BjJekrDXz:b-jeojcn) ([raw file](https://github.com/eastgenomics/gene_annotation2bed/blob/5ec22e249b79ec4b719ed0ee1cb90484fbb748ab/tests/test_gene_annotation2bed.py#L208)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

@RSWilson1 I still think this wants changing to use subTest or similar, especially where there's assertions in a loop
</blockquote></details>

Yep converted it to subTest now


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
PEP8, change line indentation

jethror1 reviewed (approved):
> Reviewed 4 of 4 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @RSWilson1)



<!-- Sent from Reviewable.io -->

jethror1 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/gene_annotation2bed/3)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @RSWilson1)



<!-- Sent from Reviewable.io -->
